db/db_impl.cc

```diff
@@ -1011,6 +1011,20 @@ void DBImpl::PurgeObsoleteFiles(const JobContext& state, bool schedule_only) {
       case kLogFile:
         keep = ((number >= state.log_number) ||
                 (number == state.prev_log_number));
+
+        /* We need to check if the log file is already added to recycle list
+         * before we mark it for deletion.
+         */
+        if (!keep && (std::find(log_recycle_files.begin(),
+            log_recycle_files.end(),number) != log_recycle_files.end())) {
```

Comment:
But, I agree if we can do it in FindObseleteFiles() , would be good as I feel mutex lock check is necessary to avoid deadlock and some extra bit of coding is required for that.
I will come up with something..